### ISSUE TITLE
Events sent in quick succession fail to send due to design issue

### ISSUE THREAD

**Issue by Cldfire** [klemens/hyper-sse#1]
~~(Posting this as a warning for any potential future passerby, I doubt this crate will see further updates.)~~

Due to the following design flaw, events sent in quick succession will see deterministic failure:

https://github.com/klemens/hyper-sse/blob/a3cec1b571260a15d08995f7b935f6f2fa80bc42/src/lib.rs#L290

`send_data()` is called without ensuring that the channel is ready to receive a chunk via `poll_ready()`. `poll_ready()` can only be called within the context of a Tokio task, something this library avoids.

Quick (dirty) fix (declare `chunk` to be mutable) (EDIT: read below comment, this code needs a timeout as well):

```rust
loop {
    result = self.tx.send_data(chunk);

    match result {
        Ok(_) => break,
        Err(c) => {
            chunk = c;
            thread::sleep(Duration::from_millis(10));
        }
    }
}
```

**Comment by klemens** [comment]
Yes, I agree that the current design won't work for sending events in rapid succession. Hyper uses a channel with only one slot here, so if you try to send only two events directly one after the other, this already fails. Additionally, the logic currently also automatically removes clients that had an error after only 5 seconds.

Just looping forever will also not work, as a stale client that has lost the connection will block your loop forever. We could introduce a timeout here to "fix" this, but I guess the real solution is to migrate to hyper 0.13 and move to async. I haven't done this yet, as I am using this with a rocket project, which also hasn't moved to async yet.

Another possible solution while staying sync would be using futures::sync::mpsc::channel directly, because then we could [differentiate](https://docs.rs/futures/0.1.29/futures/sync/mpsc/struct.TrySendError.html) between a full channel and a disconnected receiver and also use a slightly larger channel.

> I doubt this crate will see further updates

How do you come to that conclusion? :wink:

**Comment by Cldfire** [comment]
> Just looping forever will also not work, as a stale client that has lost the connection will block your loop forever. We could introduce a timeout here to "fix" this, but I guess the real solution is to migrate to hyper 0.13 and move to async. I haven't done this yet, as I am using this with a rocket project, which also hasn't moved to async yet.

Ah, good point. I needed a quick fix to move on with my own code and didn't dig that deep.

Yeah, it would appear the best solution is to migrate to a full-async stack. I agree, though, that this is a tough sell at the current point in time; the ecosystem is still a bit of a mess when it comes to the async side of things.

> Another possible solution while staying sync would be using futures::sync::mpsc::channel directly, because then we could [differentiate](https://docs.rs/futures/0.1.29/futures/sync/mpsc/struct.TrySendError.html) between a full channel and a disconnected receiver and also use a slightly larger channel.

Good thought. If I have time to poke at that in the future I may.

> > I doubt this crate will see further updates
> 
> How do you come to that conclusion? wink

Hah, my apologies! I assumed based on the lack of activity that this wasn't widely used and therefore wasn't going to be maintained. I stand corrected :stuck_out_tongue:

**Comment by klemens** [comment]
I have played around with this a bit, and I think I got it working properly:

I first converted everything to async and hyper 0.13 and then introduced a sync adapter using a normal, non-async channel and an async task that waits on that channel on its own thread:

```rust
let adapter_task = async {
    loop {
        // This will create an additional thread that just
        // waits on the non-async channel and forwards
        // everything into the async world.
        let (channel, chunk) = tokio::task::spawn_blocking(move || {
            self.sync_channel_receiver.recv().unwrap()
        }).await.unwrap();
        self.send_chunk_to_channel(chunk, channel).await;
    }
};
```

This seems to work fine and doesn't lose messages anymore. I was also able to run hyper single-threaded, so we only need two threads in the end. It seems the spawn_blocking tread is also reused and not created fresh every time.

I will push this, once I have cleaned it up bit, it is currently missing all error handling. :wink:

**Comment by Cldfire** [comment]
Awesome, sounds good!
